src/runtime/virtcontainers/utils/utils.go

```diff
@@ -334,49 +375,15 @@ func WaitLocalProcess(pid int, timeoutSecs uint, initialSignal syscall.Signal, l
 		}
 	}
 
-	pidRunning := true
-
-	secs := time.Duration(timeoutSecs)
-	timeout := time.After(secs * time.Second)
-
-	// Wait for the VM process to terminate
-outer:
-	for {
-		select {
-		case <-time.After(50 * time.Millisecond):
-			// Check if the process is running periodically to avoid a busy loop
-
-			var _status syscall.WaitStatus
-			var _rusage syscall.Rusage
-			var waitedPid int
-
-			// "A watched pot never boils" and an unwaited-for process never appears to die!
-			waitedPid, err = syscall.Wait4(pid, &_status, syscall.WNOHANG, &_rusage)
-
-			if waitedPid == pid && err == nil {
-				pidRunning = false
-				break outer
-			}
-
-			if err = syscall.Kill(pid, syscall.Signal(0)); err != nil {
-				pidRunning = false
-				break outer
-			}
-
```

Comment:
@alex-matei since the purpose of this patch is to prevent a potentially unneeded 50ms delay, what about doing just that and do the checks before the `select {}` ?

Reply:
Good point, I changed their order.

Reply:
Err... sorry if I wasn't clear. My suggestion is to do just that simple change to the existing code instead of a massive rewrite.

Reply:
It doesn't completely solve the issue if the hypervisor quits after 10 ms for example. You will still wait for 40 ms.

Reply:
> It doesn't completely solve the issue if the hypervisor quits after 10 ms for example. You will still wait for 40 ms more than needed.

True. I've been thinking a bit harder and I agree it would be good to switch to pidfd as you're proposing, since it fully addresses the issue. But I'd really like this commit to be broken down in smaller pieces, for easier review and overall maintainability. This is generally the first reason for me pushing back :smiley:

I'll add some more comments.